### PR TITLE
udev: enable Delegate=

### LINKED ISSUE #16867: udevd killing udevadm reload command when SELinux is enabled
**systemd version the issue has been seen with**
> systemd 246 (246.2-1)
+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +ZSTD +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=hybrid
   

**Used distribution**
 > Debian sid

**Expected behaviour you didn't see**
 > Normal udevd reload, like on a SELinux disabled system:

```
Aug 27 10:04:59 debianHome systemd[1]: Reloading Rule-based Manager for Device Events and Files.
Aug 27 10:04:59 debianHome udevadm[57886]: SELinux enabled state cached to: disabled
Aug 27 10:04:59 debianHome systemd-udevd[57881]: Received udev control message (RELOAD)
Aug 27 10:04:59 debianHome systemd-udevd[57881]: Unload module index
Aug 27 10:04:59 debianHome systemd-udevd[57881]: Unloaded link configuration context.
Aug 27 10:04:59 debianHome systemd[1]: Reloaded Rule-based Manager for Device Events and Files.
```
**Unexpected behaviour you saw**
 > udevd killing (or trying to kill) the udevadm process, called by the ExecReload= command:

> Enforcing mode, i.e. SIGKILL denied
```
Aug 27 10:07:14 rpi3 systemd[1]: Reloading Rule-based Manager for Device Events and Files.
Aug 27 10:07:14 rpi3 udevadm[26973]: SELinux enabled state cached to: enabled
Aug 27 10:07:14 rpi3 udevadm[26973]: Successfully loaded SELinux database in 12.371ms, size on heap is 2355K.
Aug 27 10:07:14 rpi3 audit[26970]: AVC avc:  denied  { sigkill } for  pid=26970 comm="systemd-udevd" scontext=system_u:system_r:udev_t:s0-s0:c0.c1023 tcontext=system_
u:system_r:udev_t:s0-s0:c0.c1023 tclass=process permissive=0
Aug 27 10:07:14 rpi3 audit[26970]: SYSCALL arch=c00000b7 syscall=129 success=no exit=-13 a0=695d a1=9 a2=695d a3=aaaacc2fa510 items=0 ppid=1 pid=26970 auid=4294967295
 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=4294967295 comm="systemd-udevd" exe="/usr/bin/udevadm" subj=system_u:system_r:udev_t:s0-s0:c0.c1023 key=(n
ull)
Aug 27 10:07:14 rpi3 audit: OBJ_PID opid=26973 oauid=-1 ouid=0 oses=-1 obj=system_u:system_r:udev_t:s0-s0:c0.c1023 ocomm="udevadm"
Aug 27 10:07:14 rpi3 audit: PROCTITLE proctitle="/lib/systemd/systemd-udevd"
Aug 27 10:07:14 rpi3 systemd-udevd[26970]: Received udev control message (RELOAD)
Aug 27 10:07:14 rpi3 audit[26970]: AVC avc:  denied  { sigkill } for  pid=26970 comm="systemd-udevd" scontext=system_u:system_r:udev_t:s0-s0:c0.c1023 tcontext=system_
u:system_r:udev_t:s0-s0:c0.c1023 tclass=process permissive=0
Aug 27 10:07:14 rpi3 audit[26970]: SYSCALL arch=c00000b7 syscall=129 success=no exit=-13 a0=695d a1=9 a2=695d a3=aaaacc2fa510 items=0 ppid=1 pid=26970 auid=4294967295
 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=4294967295 comm="systemd-udevd" exe="/usr/bin/udevadm" subj=system_u:system_r:udev_t:s0-s0:c0.c1023 key=(n
ull)
Aug 27 10:07:14 rpi3 audit: OBJ_PID opid=26973 oauid=-1 ouid=0 oses=-1 obj=system_u:system_r:udev_t:s0-s0:c0.c1023 ocomm="udevadm"
Aug 27 10:07:14 rpi3 audit: PROCTITLE proctitle="/lib/systemd/systemd-udevd"
Aug 27 10:07:14 rpi3 audit[26970]: AVC avc:  denied  { sigkill } for  pid=26970 comm="systemd-udevd" scontext=system_u:system_r:udev_t:s0-s0:c0.c1023 tcontext=system_
u:system_r:udev_t:s0-s0:c0.c1023 tclass=process permissive=0
Aug 27 10:07:14 rpi3 audit[26970]: SYSCALL arch=c00000b7 syscall=129 success=no exit=-13 a0=695d a1=9 a2=695d a3=aaaacc2fa510 items=0 ppid=1 pid=26970 auid=4294967295
 uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=4294967295 comm="systemd-udevd" exe="/usr/bin/udevadm" subj=system_u:system_r:udev_t:s0-s0:c0.c1023 key=(n
ull)
Aug 27 10:07:14 rpi3 audit: OBJ_PID opid=26973 oauid=-1 ouid=0 oses=-1 obj=system_u:system_r:udev_t:s0-s0:c0.c1023 ocomm="udevadm"
Aug 27 10:07:14 rpi3 audit: PROCTITLE proctitle="/lib/systemd/systemd-udevd"
Aug 27 10:07:14 rpi3 systemd-udevd[26970]: Unload module index
Aug 27 10:07:14 rpi3 systemd-udevd[26970]: Unloaded link configuration context.
Aug 27 10:07:14 rpi3 systemd[1]: Reloaded Rule-based Manager for Device Events and Files.
```

> Permissive mode, i.e. SIGKILL allowed
```
Aug 27 10:07:46 rpi3 audit[26979]: SYSCALL arch=c00000b7 syscall=64 success=yes exit=1 a0=3 a1=ffffeaff79b0 a2=1 a3=ffff7fa43670 items=0 ppid=26291 pid=26979 auid=0 u
id=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=pts1 ses=9 comm="setenforce" exe="/usr/sbin/setenforce" subj=root:sysadm_r:sysadm_t:s0-s0:c0.c1023 key=(null)
Aug 27 10:07:46 rpi3 audit: PROCTITLE proctitle=736574656E666F7263650030
Aug 27 10:07:49 rpi3 systemd[1]: Reloading Rule-based Manager for Device Events and Files.
Aug 27 10:07:49 rpi3 udevadm[26982]: SELinux enabled state cached to: enabled
Aug 27 10:07:49 rpi3 udevadm[26982]: Successfully loaded SELinux database in 12.344ms, size on heap is 2355K.
Aug 27 10:07:49 rpi3 audit[26970]: AVC avc:  denied  { sigkill } for  pid=26970 comm="systemd-udevd" scontext=system_u:system_r:udev_t:s0-s0:c0.c1023 tcontext=system_
u:system_r:udev_t:s0-s0:c0.c1023 tclass=process permissive=1
Aug 27 10:07:49 rpi3 audit[26970]: SYSCALL arch=c00000b7 syscall=129 success=yes exit=0 a0=6966 a1=9 a2=6966 a3=aaaacc2fa510 items=0 ppid=1 pid=26970 auid=4294967295 
uid=0 gid=0 euid=0 suid=0 fsuid=0 egid=0 sgid=0 fsgid=0 tty=(none) ses=4294967295 comm="systemd-udevd" exe="/usr/bin/udevadm" subj=system_u:system_r:udev_t:s0-s0:c0.c1023 key=(nu
ll)
Aug 27 10:07:49 rpi3 audit: OBJ_PID opid=26982 oauid=-1 ouid=0 oses=-1 obj=system_u:system_r:udev_t:s0-s0:c0.c1023 ocomm="udevadm"
Aug 27 10:07:49 rpi3 audit: PROCTITLE proctitle="/lib/systemd/systemd-udevd"
Aug 27 10:07:49 rpi3 systemd-udevd[26970]: Received udev control message (RELOAD)
Aug 27 10:07:49 rpi3 systemd[1]: systemd-udevd.service: Control process exited, code=killed, status=9/KILL
Aug 27 10:07:49 rpi3 systemd[1]: Reload failed for Rule-based Manager for Device Events and Files.
```

Now systemd (pid1) reports the reload as failed, because the ExecReload= process got killed.

**Steps to reproduce the problem**

```
systemctl reload systemd-udevd
```

### LINKED ISSUE #22686: Udevadm will fail to reload continuously for multiple times
**systemd version the issue has been seen with**
 > 250/main

**Used distribution**
 > …

**Linux kernel version used** (`uname -a`)

 > linux 5.10

**CPU architecture issue was seen on**
 > …

**Expected behaviour you didn't see**
 >
systemctl reload systemd-udevd.service
systemctl reload systemd-udevd.service
systemctl reload systemd-udevd.service  
no failed

**Unexpected behaviour you saw**
 > 
systemctl reload systemd-udevd.service
systemctl reload systemd-udevd.service
systemctl reload systemd-udevd.service  
failed

**Steps to reproduce the problem**
 > 
systemctl reload systemd-udevd.service
systemctl reload systemd-udevd.service
systemctl reload systemd-udevd.service  

**Additional program output to the terminal or log subsystem illustrating the issue**
https://github.com/systemd/systemd/pull/21762 fix udev reload failed problem, but the latest version of main ，Udevd will fail to reload continuously for multiple times。

```
this is ok , even try multiple times
systemctl stop systemd-udevd-control.socket
systemctl stop systemd-udevd-kernel.socket
systemctl restart systemd-udevd.service
systemctl reload systemd-udevd.service
systemctl status systemd-udevd.service
```
```
this is not ok , when try multiple times。but the time interval becomes longer， it will ok
systemctl reload systemd-udevd.service
systemctl reload systemd-udevd.service
...
```
log
Mar 09 15:48:40 localhost.localdomain systemd[1]: systemd-udevd.service: Control process exited, code=killed, status=9/KILL
Mar 09 15:48:40 localhost.localdomain systemd[1]: Reload failed for Rule-based Manager for Device Events and Files.
Mar 09 16:02:32 localhost.localdomain systemd[1]: Reloading Rule-based Manager for Device Events and Files...
Mar 09 16:02:32 localhost.localdomain systemd[1]: Reloaded Rule-based Manager for Device Events and Files.
Mar 09 16:03:26 localhost.localdomain systemd[1]: Reloading Rule-based Manager for Device Events and Files...
Mar 09 16:03:26 localhost.localdomain systemd[1]: systemd-udevd.service: Control process exited, code=killed, status=9/KILL
Mar 09 16:03:26 localhost.localdomain systemd[1]: Reload failed for Rule-based Manager for Device Events and Files.
Mar 09 16:04:40 localhost.localdomain systemd[1]: Reloading Rule-based Manager for Device Events and Files...
Mar 09 16:04:40 localhost.localdomain systemd[1]: systemd-udevd.service: Control process exited, code=killed, status=9/KILL
Mar 09 16:04:40 localhost.localdomain systemd[1]: Reload failed for Rule-based Manager for Device Events and Files.
```
<!-- Please paste relevant program terminal or journal output here, ideally
     when generated in debug mode (try setting the SYSTEMD_LOG_LEVEL=debug
     environment variable). For very long copy/pasted data consider using a
     service like https://gist.github.com/. Where copy/paste is not possible
     (for example early boot or late shutdown), a photo of the screen might do
     too, but text is always much preferred. -->
```text
…
```

### PR DESCRIPTION
**Pull request by yuwata**
This replaces ccadf9ac0d6d206767294b3f96f41eb42b48d1b0.

Fixes #16867.
Fixes #22686.

### CODE REVIEW

File: src/udev/udev-ctrl.c [hunk 1]
```diff
@@ -186,19 +203,23 @@ static int udev_ctrl_connection_event_handler(sd_event_source *s, int fd, uint32
...
-
-        cred = (struct ucred *) CMSG_DATA(cmsg);
-
         if (cred->uid != 0) {
-                log_error("Invalid sender uid "UID_FMT", ignoring message", cred->uid);
+                log_warning("Invalid sender UID "UID_FMT", ignoring message.", cred->uid);
                 return 0;
         }
+        if (cred->pid <= 0) {
+                log_warning("Invalid sender PID. ignoring message.");
```
Comment (marcosfrm) [inline src/udev/udev-ctrl.c]: Comma

Comment (yuwata) [reply src/udev/udev-ctrl.c]: Ouch. Thanks!

**Comment by poettering** [comment]
Hmm, shouldn't the clean-up of the pids set work the other way round? i.e. enumeriate processes in the cgroup and only retain the pids we actually find?

**Comment by poettering** [comment]
of course, PIDs suck as identifiers to use for this, given how quickly they are recycled.

Maybe a different approach would be this: instead of filtering by PID filter by binary. Specifically, implement a manual kill loop, that iterates through all processes in the cgroup, and kills them, but first checks with `stat_inode_same()` that `/proc/$PID/exe` refers to the same inode as `/usr/bin/udevadm`?

**Comment by poettering** [comment]
or we could use pidfds for this. i.e. instead of storing pids, store pidfds of the client process. That's kinda nice, because it pins the PID, and will signal process exit via POLLIN on the fd, which we could use to clean up our set of fds to watch.

**Comment by poettering** [comment]
Here's another idea: turn on delegation in udevd.service. run te workers in a subcgroup, and kill that one. I think that's actually the best approach.

**Comment by yuwata** [comment]
> Here's another idea: turn on delegation in udevd.service. run te workers in a subcgroup, and kill that one. I think that's actually the best approach.

Using Delegate= sounds nice. But it is not necessary to run workers in a subcgroup, as when Delegate=yes, udevadm invoked through ExecReload= is already assigned to .control subcgroup and not killed by cg_kill().

PTAL.

**Comment by poettering** [comment]
hmm, on cgroupv2 inner cgroups cannot have processes. thus udev also would need to move itself down into its own cgroup if delegation is on, if ExecREload= is used. See the second to last paragaph discussed in the "Delegation" section of https://systemd.io/CGROUP_DELEGATION

**Comment by poettering** [comment]
So, to be fully compatible with cgroupsv2 I guess the main function of udev should very early on do something like this from a new helper function:

```c
r = cg_pid_get_path(SYSTEMD_CGROUP_CONTROLLER, 0, &our_cgroup);
…
r = cg_get_xattr_malloc(SYSTEMD_CGROUP_CONTROLLER, our_cgroup, "trusted.delegate", &s);
…
r = parse_boolean(s);
if (r > 0) { /* We are invoked with our own delegated cgroup tree, let's move us one level down, so that we don't collide with the "no processes in inner nodes" rule of cgroups, when the service manager invokes the ExecReload= job in the .control/ subcgroup. */
        j = path_join(our_cgroup, "/udev");
        …
        r = cg_create_and_attach(SYSTEMD_CGROUP_CONTROLLER, j, 0);
        …
}
```

**Comment by poettering** [comment]
(background: we set the `trusted.delegate` xattr to "1" on cgroups where delegation is enabled. We can thus check it to figure out if we should do our cgroup handling and thus move ourselves down. This is useful so that we retain compat with environments that call udev from the cmdline for debugging for example)

**Comment by poettering** [comment]
But I love the simplicity of this approach, even if we need to do that cgroup migration one level down first.

**Comment by yuwata** [comment]
@poettering Thank you for the suggestion. I hope now your comments are addressed. PTAL.

**Comment by poettering** [comment]
Excellent. THis is great!

**Comment by yuwata** [comment]
Setting `Delegate=yes` to udevd.service breaks TEST-19. Now, I switched it to `Delegate=pids devices bpf-devices`. Is this a right way to delegate? @poettering PTAL.

**Comment by yuwata** [comment]
Now, Delegate=pids is used for udevd.service, as it is not necessary to enable devices or bpf-devices controllers. They are used by PID1, but not used by the service process.

**Pull request merged**